packages/xo-server/src/xo-mixins/metadata-backups.js

```diff
@@ -261,4 +311,212 @@ export default class metadataBackup {
+      const backups = backupsByPool[poolId]
+      if (backups.length === 0) {
+        delete backupsByPool[poolId]
+      } else {
+        backups.sort(compareTimestamp)
+      }
+    })
+
+    return backupsByPool
+  }
+
+  //  {
+  //    xo: {
+  //      [remote ID]: xoBackups
+  //    },
+  //    pool: {
+  //      [remote ID]: poolBackups
+  //    }
+  //  }
+  async listMetadataBackups(remoteIds: string[]) {
+    const app = this._app
+
+    const xo = {}
+    const pool = {}
+    await Promise.all(
+      remoteIds.map(async remoteId => {
+        try {
+          const handler = await app.getRemoteHandler(remoteId)
+
+          const [xoList, poolList] = await Promise.all([
+            this._listXoMetadataBackups(remoteId, handler),
+            this._listPoolMetadataBackups(remoteId, handler),
+          ])
+          if (xoList.length !== 0) {
+            xo[remoteId] = xoList
+          }
+          if (Object.keys(poolList).length !== 0) {
+            pool[remoteId] = poolList
+          }
+        } catch (error) {
+          log.warn(`listMetadataBackups for remote ${remoteId}`, { error })
+        }
+      })
+    )
+
+    return {
+      xo,
+      pool,
+    }
+  }
+
+  // Task logs emitted in a restore execution:
+  //
+  // task.start(message: 'restore', data: <Metadata />)
+  // └─ task.end
+  async restoreMetadataBackup(id: string) {
+    const app = this._app
+    const logger = this._logger
+    const message = 'metadataRestore'
+    const [remoteId, dir, ...path] = id.split('/')
+
+    let taskId
+    try {
+      const handler = await app.getRemoteHandler(remoteId)
+
+      const metadataFolder = `${dir}/${path.join('/')}`
+
+      taskId = logger.notice(message, {
+        event: 'task.start',
+        data: JSON.parse(
+          String(await handler.readFile(`${metadataFolder}/metadata.json`))
+        ),
+      })
+      this._runningMetadataRestores.add(taskId)
+
+      let result
+      if (dir === DIR_XO_CONFIG_BACKUPS) {
+        result = await app.importConfig(
+          JSON.parse(
+            String(await handler.readFile(`${metadataFolder}/data.json`))
+          )
+        )
+      } else {
+        result = await app
+          .getXapi(path[1])
+          .importPoolMetadata(
+            await handler.createReadStream(`${metadataFolder}/data`),
+            true
+          )
+      }
+
+      logger.notice(message, {
+        event: 'task.end',
+        result,
+        status: 'success',
+        taskId,
+      })
+    } catch (error) {
+      this._runningMetadataRestores.delete(taskId)
+      logger.error(message, {
+        event: 'task.end',
+        result: serializeError(error),
+        status: 'failure',
+        taskId,
+      })
+      throw error
+    }
+  }
+
+  async deleteMetadataBackup(id: string) {
+    const uuidReg = '\\w{8}(-\\w{4}){3}-\\w{12}'
+    const metadataDirReg = 'xo-(config|pool-metadata)-backups'
+    const timestampReg = '\\d{8}T\\d{6}Z'
```

Comment:
This does not work for pool metadata backups due to the extra directory level.

I think it would be much easier to simply read the metadata file.

Reply:
This case is handled by the [regular expression](https://github.com/vatesfr/xen-orchestra/pull/4096/files/d5ad781adf6a2c409d7648df367a169b4135af17#diff-fe664de5297cdcda5f09b752eeab1fd6R510)